service/controller/release/resource/app/desired.go

```diff
@@ -0,0 +1,121 @@
+package app
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	applicationv1alpha1 "github.com/giantswarm/apiextensions/pkg/apis/application/v1alpha1"
+	releasev1alpha1 "github.com/giantswarm/apiextensions/pkg/apis/release/v1alpha1"
+	"github.com/giantswarm/microerror"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/giantswarm/release-operator/pkg/project"
+	"github.com/giantswarm/release-operator/service/controller/key"
+)
+
+func (r *resourceStateGetter) GetDesiredState(ctx context.Context, obj interface{}) ([]*applicationv1alpha1.App, error) {
+	cr, err := key.ToReleaseCR(obj)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	var desiredComponents []releasev1alpha1.ReleaseSpecComponent
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("finding desired components for release %#q", cr.Name))
+
+		desiredComponents, err = r.getDesiredComponents(ctx, cr)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found %d desired components for release %#q", len(desiredComponents), cr.Name))
+	}
+
+	var desiredAppCRs []*applicationv1alpha1.App
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("computing desired app CRs for release %#q components", cr.Name))
+
+		for _, c := range desiredComponents {
+			appCR := newAppCR(appCRName(c), c.Name, c.Version, r.appCatalog)
+			desiredAppCRs = append(desiredAppCRs, appCR)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("computed %d desired app CRs for release %#q components", len(desiredAppCRs), cr.Name))
+	}
+
+	return desiredAppCRs, nil
+}
+
+func (r *resourceStateGetter) getDesiredComponents(ctx context.Context, cr *releasev1alpha1.Release) ([]releasev1alpha1.ReleaseSpecComponent, error) {
+	// If this is a non-EOL release and it isn't deletion event all
+	// components of that release are desired.
+	cr.GetDeletionTimestamp()
+	if cr.Status.Cycle.Phase != releasev1alpha1.CyclePhaseEOL && !key.IsDeleted(cr) {
+		return cr.Spec.Components, nil
+	}
+
+	var activeReleases []releasev1alpha1.Release
+	{
+		opts := metav1.ListOptions{
+			LabelSelector: key.LabelReleaseCyclePhase + "!=" + releasev1alpha1.CyclePhaseEOL.String(),
```

Comment:
Is there another more standard way of computing the label selector?

Reply:
I don't know. Everything I saw so far was string concatenation or formatting.